### PR TITLE
A simple CO2 simulation example based on PTFlash

### PR DESCRIPTION
**Pull request by GitPaean**
It was developed by @trinemykk . and I did various fixing and cleaning up. 

There are still more rooms for improvements, especially for long term, while the development is relatively isolated from other main development activities, and this PR is not modifying any existing code. 

Creating this PR now of reviewing and potentially merging.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Review by totto82 — COMMENTED** [review]
Good work. I think this is almost merge ready. I only have some small comments / suggestions.

File: opm/models/ptflash/flashintensivequantities.hh [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        const Scalar R = Opm::Constants<Scalar>::R;
+        Evaluation Z_L = (paramCache.molarVolume(FluidSystem::oilPhaseIdx) * fluidState_.pressure(FluidSystem::oilPhaseIdx) )/
+        (R * fluidState_.temperature(FluidSystem::oilPhaseIdx));
+        paramCache.updatePhase(fluidState_, FluidSystem::gasPhaseIdx);
+        Evaluation Z_V = (paramCache.molarVolume(FluidSystem::gasPhaseIdx) * fluidState_.pressure(FluidSystem::gasPhaseIdx) )/
+        (R * fluidState_.temperature(FluidSystem::gasPhaseIdx));
+
+
+        // Update saturation
+        Evaluation L = fluidState_.L();
```
[the remark this answers is not in the document]
Comment (GitPaean) [reply opm/models/ptflash/flashintensivequantities.hh:208]: Done!

File: opm/models/ptflash/flashmodel.hh [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+        unsigned compIdx = pvIdx - Indices::conti0EqIdx;
+
+        // make all kg equal. also, divide the weight of all total
+        // compositions by 100 to make the relative errors more
+        // comparable to the ones of the other models (at 10% porosity
+        // the medium is fully saturated with water at atmospheric
+        // conditions if 100 kg/m^3 are present!)
+        return FluidSystem::molarMass(compIdx) / 100.0;
+    }
```
[the remark this answers is not in the document]
Comment (GitPaean) [reply opm/models/ptflash/flashmodel.hh:304]: It is from the old flashmodel.hh under the folder flash.

[the remark this answers is not in the document]
Comment (GitPaean) [reply opm/models/ptflash/flashmodel.hh:304]: I do not think it is used in the ptflash

File: tests/problems/co2ptflashproblem.hh [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+struct CellsX<TypeTag, TTag::CO2PTBaseProblem> { static constexpr int value = 30; };
+template<class TypeTag>
+struct CellsY<TypeTag, TTag::CO2PTBaseProblem> { static constexpr int value = 1; };
+// CellsZ is not needed, while to keep structuredgridvanguard.hh compile
+template<class TypeTag>
+struct CellsZ<TypeTag, TTag::CO2PTBaseProblem> { static constexpr int value = 1; };
+
+
+// compositional, with diffusion
+template <class TypeTag>
```
[the remark this answers is not in the document]
Comment (GitPaean) [reply tests/problems/co2ptflashproblem.hh:283]: Indeed!

**Comment by GitPaean** [comment]
Thanks for the comments. Will finalize the update later.

File: opm/models/ptflash/flashmodel.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (GitPaean) [reply opm/models/ptflash/flashmodel.hh:304]: removed.

File: opm/models/ptflash/flashnewtonmethod.hh [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        nextValue -= update;
+
+        ////
+        // Pressure updates
+        ////
+        // limit pressure reference change to 20% of the total value per iteration
+        clampValue_(nextValue[pressure0Idx],
+                    currentValue[pressure0Idx]*0.8,
+                    currentValue[pressure0Idx]*1.2);
+
```
[the remark this answers is not in the document]
Comment (GitPaean) [reply opm/models/ptflash/flashnewtonmethod.hh:97]: reformulating a little bit to make it easier in the future to make it a running parameter.  At this stage, not sure whether we want to tune it directly.

File: tests/problems/co2ptflashproblem.hh [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+    Scalar porosity([[maybe_unused]] const Context& context, [[maybe_unused]] unsigned spaceIdx, [[maybe_unused]] unsigned timeIdx) const
+    {
+        int spatialIdx = context.globalSpaceIndex(spaceIdx, timeIdx);
+        int inj = 0;
+        int prod = EWOMS_GET_PARAM(TypeTag, unsigned, CellsX) - 1;
+        if (spatialIdx == inj || spatialIdx == prod) {
+            return 1.0;
+        } else {
+            return porosity_;
+        }
```
[the remark this answers is not in the document]
Comment (GitPaean) [reply tests/problems/co2ptflashproblem.hh:499]: We did not adjust the perm for the well cell, so they are 100MD. And also, we do not have wells yet here.

File: tests/problems/co2ptflashproblem.hh [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+            comp[1] = Evaluation::createVariable(0.01 - 1e-3, 2);
+            comp[2] = 1. - comp[0] - comp[1];
+        }
+        ComponentVector sat;
+        sat[0] = 1.0;
+        sat[1] = 1.0 - sat[0];
+        const Scalar temp = 423.25;
+
+        Scalar p0 = 75e5; // CONVERGENCE FAILURE WITH 75
+
```
[the remark this answers is not in the document]
Comment (GitPaean) [reply tests/problems/co2ptflashproblem.hh:562]: Temperature are the same cross the domain and the pressure are used while we make the leftmost cell has higher pressure (150barsa) and rightmost cell has lower pressure(37.5barsa), compared with the rest of the cells (75 barsa).

**Comment by GitPaean** [comment]
jenkins build this please

File: tests/problems/co2ptflashproblem.hh (hunk 6 above)
[the remark this answers is not in the document]
Comment (totto82) [reply tests/problems/co2ptflashproblem.hh:562]: I was thinking like this. 
`Scalar p0 = EWOMS_GET_PARAM(TypeTag, Scalar, Initialpressure);`
`Scalar temp = EWOMS_GET_PARAM(TypeTag, Scalar, Temperature);
Currently the input parameter `initialPressure `and `Temperature `are not used. `

[the remark this answers is not in the document]
Comment (GitPaean) [reply tests/problems/co2ptflashproblem.hh:560]: temperature is used, not thinking it is the best way.  I will remove `const Scalar temp = 423.25;` now. 
```c++
    void initPetrophysics()
    {
        temperature_ = EWOMS_GET_PARAM(TypeTag, Scalar, Temperature);
        K_ = this->toDimMatrix_(9.869232667160131e-14);

        porosity_ = 0.1;
    }
```

Will update to use `Scalar p0 = EWOMS_GET_PARAM(TypeTag, Scalar, Initialpressure);`. 

We probably can improve how the way it works along the way.

**Comment by GitPaean** [comment]
jenkins build this please

**Pull request merged**